src/selectors/AutoTuner.h

```diff
@@ -0,0 +1,124 @@
+      : _tuningInterval(tuningInterval),
+        _iterationsSinceTuning(tuningInterval),  // init to max so that tuning happens in first iteration
+        _containerSelector(boxMin, boxMax, cutoff, allowedContainerOptions, allowedTraversalOptions) {}
+
+  /**
+   * Getter for the optimal container.
+   * @return Smartpointer to the optimal container.
+   */
+  std::shared_ptr<autopas::ParticleContainer<Particle, ParticleCell>> getContainer() {
+    return _containerSelector.getOptimalContainer();
+  };
+
+  /**
+   * Function to iterate over all pairs of particles in the container.
+   * This function only handles short-range interactions.
+   * @tparam ParticleFunctor
+   * @param f Functor that describes the pair-potential
+   * @param useSoA if true SoA data structure is used otherwise AoS
+   */
+  template <class ParticleFunctor>
+  void iteratePairwise(ParticleFunctor *f, bool useSoA);
+
+ private:
+  template <class CellFunctor>
+  void tune(CellFunctor &cellFunctor);
+
+  unsigned int _tuningInterval, _iterationsSinceTuning;
+  ContainerSelector<Particle, ParticleCell> _containerSelector;
+};
+template <class Particle, class ParticleCell>
+template <class ParticleFunctor>
+void AutoTuner<Particle, ParticleCell>::iteratePairwise(ParticleFunctor *f, bool useSoA) {
+  bool newton3Allowed = f->allowsNewton3();
+  bool nonNewton3Allowed = f->allowsNonNewton3();
+  bool useNewton3 = false;
+  if (newton3Allowed and nonNewton3Allowed) {
+    /// @todo auto-tune (far off future)
+  } else if (not newton3Allowed and not nonNewton3Allowed) {
+    /// @todo throw exception
+  } else {
+    useNewton3 = newton3Allowed;
+  }
+
+  /// @todo: WANT one single iteratePairwise(CellFunctor) for containers
+  /// @todo: CellFunctor for iteration should be build here using selectors for SoA and N3
+  if (useSoA) {
+    if (useNewton3) {
+      CellFunctor<Particle, ParticleCell, ParticleFunctor, true, true> cellFunctor(f);
+      //      if (_iterationsSinceTuning == _tuningInterval)
+      //        tune(cellFunctor);
+      //      _containerSelector.getOptimalContainer()->iteratePairwiseSoA(cellFunctor);
+      _containerSelector.getOptimalContainer()->iteratePairwiseSoA(f, useNewton3);
+    } else {
+      CellFunctor<Particle, ParticleCell, ParticleFunctor, true, false> cellFunctor(f);
+      //      if (_iterationsSinceTuning == _tuningInterval)
+      //        tune(cellFunctor);
+      //      _containerSelector.getOptimalContainer()->iteratePairwiseSoA(cellFunctor);
+      _containerSelector.getOptimalContainer()->iteratePairwiseSoA(f, useNewton3);
+    }
+  } else {
+    if (useNewton3) {
+      CellFunctor<Particle, ParticleCell, ParticleFunctor, false, true> cellFunctor(f);
+      //      if (_iterationsSinceTuning == _tuningInterval)
+      //        tune(cellFunctor);
+      //      _containerSelector.getOptimalContainer()->iteratePairwiseAoS(cellFunctor);
+      _containerSelector.getOptimalContainer()->iteratePairwiseAoS(f, useNewton3);
```

Comment:
In particular, I think it would make sense to create the CellFunctor here and then call iteratePairwise with it. This would remove the need for an AoS and SoA version. Also, a selector/tuner for Newton3 and Data structure could then be plugged in here. The cell functor could then be thrown at a decision tree like you suggested in #52 .
@SteffenSeckler opinions?

Reply:
#52 is now solved by #64 so please don't do the work twice :) but just use the macro.